libqtile/backend/x11/core.py

```diff
@@ -804,19 +802,15 @@ def flush(self):
     def graceful_shutdown(self):
         """Try to close windows gracefully before exiting"""
 
-        def get_interesting_pid(win):
-            # We don't need to kill Internal or Static windows, they're qtile
-            # managed and don't have any state.
-            if not isinstance(win, base.Window):
-                return None
-            try:
-                return win.window.get_net_wm_pid()
-            except Exception:
-                logger.exception("Got an exception in getting the window pid")
-                return None
-
-        pids = map(get_interesting_pid, self.qtile.windows_map.values())
-        pids = list(filter(lambda x: x is not None, pids))
+        try:
+            pids = [
+                pid
+                for win in self.qtile.windows_map.values()
+                if not isinstance(win, base.Internal) and (pid := win.get_pid())
+            ]
+        except xcffib.ConnectionException:
+            logger.warning("Server disconnected, couldn't close windows gracefully.")
+            return
```

Comment:
Would `pids` be empty at this point? If there are some processes, shouldn't we still try to close them rather than returning here?

Reply:
It's the `win.get_pid()` that tries to communication with the X server and fails, so `pids` doesn't actually get set. This is also why there are two for loops - get all pids (if we can), then kill them below. Either they will all fail or none (because it's same connection). If they fail, we don't know which pids were X clients so we can't kill any processes. From what I can tell, this situation comes about if the X server closes connections before us, and the same thing will happen to those clients, so they're all getting connection errors too, and will close themselves.